content/blog/2016/2016-04-07-pipeline-for-runs-on-hardware.adoc

```diff
@@ -0,0 +1,258 @@
+** link:https://wiki.jenkins-ci.org/display/JENKINS/Build+Failure+Analyzer[Build Failure Analyzer] allows to identify the root cause of a build failure (e.g. by build log parsing).
+** link:https://wiki.jenkins-ci.org/display/JENKINS/Conditional+BuildStep+Plugin[Conditional Build Step] and 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Flexible+Publish+Plugin[Flexible Publish] plugins allow altering the build flow according to the analysis results.
+** Combination of the plugins above is possible, but it makes job configurations extremely large.
+* Tests on hardware peripherals may take much time. 
+If an infrastructure fails, we may have to restart the run from scratch. 
+So the builds should be robust against infrastructure issues including network failures and Jenkins master restarts.
+* Tests on hardware should be reproducible, so the environment and input parameters should be controlled well.
+** Jenkins supports
+link:https://wiki.jenkins-ci.org/display/JENKINS/Workspace+Cleanup+Plugin[cleaning workspaces], so it can get rid of temporary files generated by previous runs.
+** Jenkins provides support of slaves connected via containers (e.g. 
+  link:https://wiki.jenkins-ci.org/display/JENKINS/Docker+Plugin[Docker]) or VMs, which allow creating clean environments for every new run.
+It's important for 3rd-party tools, which may modify files outside the workspace: user home directory, temporary files, etc.
+** These environments still need to be connected to hardware peripherals, which may be a serious obstacle for Jenkins admins
+
+Usage of 
+link:https://jenkins.io/doc/pipeline/[Jenkins Pipeline]  can effectively address the use-cases above.
+In this blogpost I want to address several cases.
+
+## Pipeline-as-Code for test runs on hardware
+
+Pipeline-as-Code is an emerging technology for representing job configurations as code.
+In Jenkins there are two most popular implementations: Pipeline and JobDSL plugins.
+JobDSL Plugin internally generates common freestyle jobs according to the script, so it's functionality is similar to the description above.
+In this post I will focus on the Pipeline plugin.
+
+Below you can find an example of Pipeline scripts, which runs tests on FPGA board. The id of this board comes from build parameters (_fpgaId_). In this script we also presume that all nodes have pre-installed tools (Xilinx ISE in this case).
+
+```groovy
+// Run on node having my_fpga label 
+node("linux && ml509") {
+  git url:"http://github.com/oleg-nenashev/pipeline_hw_samples";
+  sh "make all";
+}
+```
+
+But such scenario could be also implemented in a Free-style project.
+What would we get from Pipeline plugin?
+
+## Getting added-value from Pipeline-as-code
+
+Pipeline provides much added-value features for hardware-based tests. 
+I would like to highlight the following advantages:
+
+* Robustness against restarts of Jenkins master.
+* Robustness against network disconnects. _sh()_ steps are based on the
```

Comment:
When referring to steps, I try to always use back-ticks so the code is formatted in a monospaced font. For example, instead of:

_sh()_

using

`sh`

I only recently added this to the styleguide to ensure consistency across documentation.